salt/utils/path.py

```diff
@@ -412,3 +417,94 @@ def os_walk(top, *args, **kwargs):
+    '''
+    if root_dir:
+        root_dir = os.path.abspath(root_dir)
+
+    def_root_dir = salt.syspaths.ROOT_DIR.rstrip(os.sep)
+    tmp_path_def_root_dir = None
+    tmp_path_root_dir = None
+
+    # When running testsuite, salt.syspaths.ROOT_DIR is often empty
+    if path == def_root_dir or path.startswith(def_root_dir + os.sep):
+        # Remove the default root dir prefix
+        tmp_path_def_root_dir = path[len(def_root_dir):]
+
+    if root_dir and (path == root_dir or
+                     path.startswith(root_dir + os.sep)):
+        # Remove the root dir prefix
+        tmp_path_root_dir = path[len(root_dir):]
+
+    if tmp_path_def_root_dir and not tmp_path_root_dir:
+        # Just the default root dir matched
+        path = tmp_path_def_root_dir
+    elif tmp_path_root_dir and not tmp_path_def_root_dir:
+        # Just the root dir matched
+        path = tmp_path_root_dir
+    elif tmp_path_def_root_dir and tmp_path_root_dir:
+        # In this case both the default root dir and the override root
+        # dir matched; this means that either
+        # def_root_dir is a substring of root_dir or vice versa
+        # We must choose the most specific path
+        if def_root_dir in root_dir:
+            path = tmp_path_root_dir
+        else:
+            path = tmp_path_def_root_dir
+    elif salt.utils.platform.is_windows() and not os.path.splitdrive(path)[0]:
+        # In windows, os.path.isabs resolves '/' to 'C:\\' or whatever
+        # the root drive is.  This elif prevents the next from being
+        # hit, so that the root_dir is prefixed in cases where the
+        # drive is not prefixed on a config option
+        pass
+    elif os.path.isabs(path):
+        # Absolute path (not default or overridden root_dir)
+        # No prepending required
+        return path
+
+    if root_dir is None:
+        # Transform to absolute path without prepending root_dir
+        return os.path.abspath(path)
+
+    # Prepending the root dir
+    return os.path.join(root_dir, path)
+
+
+def expand_path(path, root_dir=None):
+    '''
+    Resolves user home '~' or prepend the 'root_dir'
+    directory and returns the expanded 'path'
+    '''
+    if path.startswith('~'):
+        path = os.path.expanduser(path)
+    elif urlparse(path).scheme == '':
+        path = prepend_root_dir(path, root_dir)
+
+    return path
+
+
+def expand_glob_path(path, root_dir=None):
+    '''
+    Resolves user home '~' or prepend the 'root_dir' directory and
+    applies shell globbing, then returns the expanded pathes.
+    '''
+    unglobbed_path = []
+
+    if path.startswith('~'):
+        path = os.path.expanduser(path)
+    elif urlparse(path).scheme == '':
+        path = prepend_root_dir(path, root_dir)
+
+    try:
+        if glob.has_magic(path) and urlparse(path).scheme == '':
+            unglobbed_path.extend(glob.glob(path))
+        else:
+            unglobbed_path.append(path)
+    except Exception:
```

Comment:
`log.debug` at least needs to be added.

Reply:
I add a `log.debug` for the `Exception` but I must admit that this code is highly inspired from `salt.config._expand_glob_path` I don't know how to make it raise without mocking.

Reply:
Don't overdo this: just if there is an exception that normally one must not see, then move it to the debug (so one will see it if he wants). But if there is an exception and you're just yanking it, then sysop won't know what is going on and there won't be a way to know it.

Reply:
I mean that in my test I'm doing:
```python
    @skipIf(NO_MOCK, NO_MOCK_REASON)
    def test_exception_returns_unglobbed_path_expanded_with_default_root_dir(self):
        with patch('glob.has_magic', side_effect=Exception):
            self.assertEqual(
                salt.utils.path.expand_glob_path(self.ABSOLUTE_GLOB_PATH),
                [salt.utils.path.join(self.DEFAULT_ROOT_DIR,
                                      self.ABSOLUTE_GLOB_PATH)])
```

So I can test the `except` branch:

``` python
    except Exception as error:
        log.debug("Unable to unglob the pattern "
                  "'{0}' expanded as '{1}': {2}".format(expanded_glob,
                                                        path,
                                                        error))
        unglobbed_path.append(expanded_glob)
```